libq/cur_sys_pkg.c

```diff
@@ -0,0 +1,348 @@
+{
+  if((int) dig > 57) 
+  {
+    return (int)dig - 87;
+  }
+  return (int )dig - 48;
+}
+
+static int compare_hash_num(char *hash1,char*hash2)
+{
+  int temp1,temp2;
+  for(int i=0;i<HASH_SIZE;++i)
+  {
+    temp1=conv_char_int(hash1[i]);
+    temp2=conv_char_int(hash2[i]);
+    if(temp2 > temp1)
+    {
+      return 1;
+    }else if (temp2 < temp1) {
+      return -1;
+    }
+  }
+  return 0;
+}
+
+static char *get_fullname_package(depend_atom *datom)
+{
+  int cat_len,name_len=0;
+  char *package_name =NULL;
+
+  assert(datom!=NULL);
+  assert(datom->CATEGORY!=NULL);
+  assert(datom->PN!=NULL);
+
+  cat_len=strlen(datom->CATEGORY);
+  name_len=strlen(datom->PN);
+  package_name=calloc((cat_len +1+ name_len +1), sizeof(*package_name));
+  package_name[cat_len + 1 +name_len]='\0';
+
+  strcat(package_name,datom->CATEGORY);
+  strcat(package_name,"/");
+  strcat(package_name,datom->PN);
+
+  return package_name;
+}
+static void add_node(cur_pkg_tree_node **root,char *data,char *key,
+                     char *package_name,unsigned int safe_to_free)
+{
+  if(*root==NULL)
+  {
+    *root=xmalloc(sizeof(**root));
+    (*root)->key=key;
+    (*root)->hash_buffer=data;
+    (*root)->package_name=package_name;
+    (*root)->safe_to_free_package_name=safe_to_free;
+    (*root)->greater=NULL;
+    (*root)->minor=NULL;
+    return;
+  }
+
+  int is_greater=compare_hash_num((*root)->key,key);
+  
+  switch (is_greater) {
+    case 0:
+          printf("there are two packages wich update the same file %s %s, the hash of the file is %s\n"
+                 ,package_name,(*root)->package_name,data);
+      return ;
+    case 1:
+      return add_node(&(*root)->greater,data,key,package_name,safe_to_free);
+    case -1:
+      return add_node(&(*root)->minor,data,key,package_name,safe_to_free);
+  }
+}
+
+static char *hash_from_file(char *file_path_complete)
+{
+  FILE *file_to_hash;
+  char buf[512];
+  unsigned char hex_hash[HASH_SIZE+1];
+  hex_hash[HASH_SIZE]='\0';
+  MD5_CTX ctx;
+  
+  char *out = NULL;
+  out=xmalloc(HASH_SIZE+1* sizeof(*out));
+  out[HASH_SIZE]='\0';
+
+  file_to_hash = fopen(file_path_complete,"r");
+  if(file_to_hash == NULL)
+  {
+    fprintf(stderr, "%s not found\n",file_path_complete);
+    out=xmalloc(3* sizeof(*out));
+    out[0]='-';
+    out[1]='1';
+    out[2]='\0';
+
+    fclose(file_to_hash);
+
+    return out;
+  }
+  MD5_Init(&ctx);
+
+  size_t byte_read=0;
+  while ( ( byte_read = fread(buf,1,512,file_to_hash) ) > 0) {
+    MD5_Update(&ctx,buf,byte_read);
+  }
+  MD5_Final(hex_hash,&ctx);
+
+  hash_hex(out,hex_hash,(HASH_SIZE>>1));
+  out[HASH_SIZE]='\0';
+
+  fclose(file_to_hash);
+
+  return out;
+}
+
+char *hash_from_string(char *str,size_t len)
+{
+  unsigned char hex_buf[HASH_SIZE+1];
+  char *hash_final=xmalloc(HASH_SIZE+1*sizeof(*hash_final));
+  hash_final[HASH_SIZE]='\0';
+  hex_buf[HASH_SIZE]='\0';
+  MD5_CTX ctx;
+  
+  MD5_Init(&ctx);
+  MD5_Update(&ctx,str,len);
+  MD5_Final(hex_buf,&ctx);
```

Comment:
it seems like a better idea to extend hash.h to add a hash_buffer or something, that would also make a fine separate commit that would be easily reviewed and merged